netket/vqs/mc/kernels.py

```diff
@@ -70,39 +104,44 @@ def local_value_kernel_flattened(logpsi: Callable, pars: PyTree, σ: Array, args
 
     return e_loc
 
-def local_value_squared_kernel_flattened(logpsi: Callable, pars: PyTree, σ: Array, args: PyTree):
+
+def local_value_squared_kernel_flattened(
+    logpsi: Callable, pars: PyTree, σ: Array, args: PyTree
+):
     """
     local_value kernel for MCState and Squared (generic) operators
     """
     return jnp.abs(local_value_kernel_flattened(logpsi, pars, σ, args)) ** 2
 
 
-@batch_discrete_kernel
-def local_value_kernel(logpsi: Callable, pars: PyTree, σ: Array, args: PyTree):
+def local_value_op_op_kernel_flattened(
+    logrho: Callable, pars: PyTree, σ: Array, args: PyTree
+):
     """
-    local_value kernel for MCState and generic operators
+    local_value kernel for MCMixedState and generic operators
     """
-    σp, mel = args
-    return jnp.sum(mel * jnp.exp(logpsi(pars, σp) - logpsi(pars, σ)))
+    η, mels, secs, N_conn = args
+    N_conn = N_conn.value
+    N_samples = σ.shape[0]
 
+    ns = jnp.diff(secs, prepend=0)
 
-def local_value_squared_kernel(logpsi: Callable, pars: PyTree, σ: Array, args: PyTree):
-    """
-    local_value kernel for MCState and Squared (generic) operators
-    """
-    return jnp.abs(local_value_kernel(logpsi, pars, σ, args)) ** 2
+    ση = jnp.repeat(σ, ns, total_repeat_length=N_conn)
+    σσ = jnp.hstack((σ, σ))
 
+    ρ_σσ = logrho(pars, σσ)
+    ρ_ση = logrho(pars, ση)
 
-@batch_discrete_kernel
-def local_value_op_op_cost(logpsi: Callable, pars: PyTree, σ: Array, args: PyTree):
-    """
-    local_value kernel for MCMixedState and generic operators
-    """
-    σp, mel = args
+    ρ_σσ_ext = jnp.repeat(ρ_σσ, ns, total_repeat_length=N_conn)
 
-    σ_σp = jax.vmap(lambda σp, σ: jnp.hstack((σp, σ)), in_axes=(0, None))(σp, σ)
-    σ_σ = jnp.hstack((σ, σ))
-    return jnp.sum(mel * jnp.exp(logpsi(pars, σ_σp) - logpsi(pars, σ_σ)))
+    delta_ρ = mels * jnp.exp(ρ_ση - ρ_σσ_ext)
+    indices = jnp.repeat(jnp.arange(N_samples), ns, total_repeat_length=N_conn)
+
+    e_loc = jax.ops.segment_sum(
```

Comment:
maybe can you write a comment on what these lines are doing?

Reply:
(This is essentially equivalent to the old C++ code that used a for loop, but is written in a jax-compatible way, which makes it rather... unreadable.)

(This PR is still WIP because I'm hoping there's a better way to implement this...)

(If we end up deciding to merge this or some future iteration, I'll of course document it. But for now I'd like some feedback on the performance gain from people like @jwnys who are working with fermonic models which have lots of zeros...)